pkg/obsservice/obslib/queue/mem_queue.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+package queue
+
+import (
+	"container/list"
+
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+	"github.com/gogo/protobuf/proto"
+)
+
+// MemoryQueue is a FIFO, in-memory event queue designed for use by the
+// observability service for buffering events awaiting aggregation.
+// It's elements are intended to be protobuf messages, which by default
+// implement the proto.Sizer interface. This provides us a convenient
+// way to determine the size of buffered events on Enqueue / Dequeue
+// to enforce the configured maxSize.
+//
+// MemoryQueue is backed by the container/list package from the Golang
+// stdlib, which is a linked list under the hood. This allows us to avoid
+// holding onto unused memory like you would by trimming a pre-allocated
+// slice (e.g. mySlice[1:]) when dequeue'ing elements from the front.
+// See: https://pkg.go.dev/container/list for details.
+//
+// See NewMemoryQueue for initialization.
+type MemoryQueue[T proto.Sizer] struct {
```

Comment:
`MemoryQueue` implementation seems to be very universal and not limited to `obsservice` only. Would it be appropriate to move it to `pkg/util`? (it already has  `pkg/util/cache` package that provides in memory cache with with specific restrictions so this one could be a one more)